e2e/testcases/workload_identity_test.go

```diff
@@ -286,3 +278,136 @@ func truncateStringByLength(s string, l int) string {
+	sourceVersion string
+	gsaEmail      string
+	rootCommitFn  nomostest.Sha1Func
+}
+
+// configureRootSync updates RootSync to sync with the provided auth.
+// It reuses the RootSync instead of creating a new one so that test resources
+// can be cleaned up after the test.
+func configureRootSync(nt *nomostest.NT, rs *v1beta1.RootSync, tenant string, sourceSpec *sourceSpec) {
+	// For helm charts, we need to push the chart to the AR before configuring the RootSync
+	if sourceSpec.sourceType == v1beta1.HelmSource {
+		chart, err := artifactregistry.PushHelmChart(nt, sourceSpec.sourceChart, sourceSpec.sourceVersion)
+		if err != nil {
+			nt.T.Fatalf("failed to push helm chart: %v", err)
+		}
+
+		sourceSpec.sourceRepo = chart.Image.RepositoryOCI()
+		sourceSpec.sourceChart = chart.Image.Name
+		sourceSpec.sourceVersion = chart.Image.Version
+		sourceSpec.rootCommitFn = nomostest.HelmChartVersionShaFn(chart.Image.Version)
+	}
+
+	nt.T.Logf("Update RootSync to sync %s from repo %s", tenant, sourceSpec.sourceRepo)
+	switch sourceSpec.sourceType {
+	case v1beta1.GitSource:
+		nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"git": {"dir": "%s", "branch": "main", "repo": "%s", "auth": "gcpserviceaccount", "gcpServiceAccountEmail": "%s", "secretRef": {"name": ""}}}}`,
+			tenant, sourceSpec.sourceRepo, sourceSpec.gsaEmail))
+	case v1beta1.OciSource:
+		nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceType": "%s", "oci": {"dir": "%s", "image": "%s", "auth": "gcpserviceaccount", "gcpServiceAccountEmail": "%s"}, "git": null}}`,
+			v1beta1.OciSource, tenant, sourceSpec.sourceRepo, sourceSpec.gsaEmail))
+	case v1beta1.HelmSource:
+		// Set the helm re-pulling duration to 5s instead of relying on the default 1h,
+		// particularly in scenarios where the new IAM policy bindings may not have
+		// been fully effective.
+		nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceType": "%s", "helm": {"chart": "%s", "repo": "%s", "version": "%s", "auth": "gcpserviceaccount", "gcpServiceAccountEmail": "%s", "releaseName": "my-coredns", "namespace": "coredns", "period": "5s"}, "git": null}}`,
+			v1beta1.HelmSource, sourceSpec.sourceChart, sourceSpec.sourceRepo, sourceSpec.sourceVersion, sourceSpec.gsaEmail))
+	}
+}
+
+func readerRole(sourceType v1beta1.SourceType, sourceRepo string) (string, error) {
+	switch sourceType {
+	case v1beta1.GitSource: // Git repo on CSR
+		return "roles/source.reader", nil
+	case v1beta1.OciSource: // OCI images on GCR
+		if strings.HasPrefix(sourceRepo, nomostesting.GCRHost) {
+			return "roles/containerregistry.ServiceAgent", nil
+		}
+		return "roles/artifactregistry.reader", nil // OCI images on AR
+	case v1beta1.HelmSource:
+		return "roles/artifactregistry.reader", nil // Helm charts on AR
+	default:
+		return "", fmt.Errorf("invalid source type %q", sourceType)
+	}
+}
+
+// grantIAMPermissions sets up IAM policy binding to grant IAM permissions
+func grantIAMPermissions(nt *nomostest.NT, principle, role string) {
```

Comment:
The test-runner GSA doesn't have permission to manage IAM currently, and I'm not sure that's something we want to add due to privilege escalation.

Can this be managed with the terraform config?

Reply:
Terraform config is a one-off configuration. The IAM permissions are created and deleted by each testcase.
I think we may need to grant the test-runner GSA permission to manage IAM.

Reply:
Why do we need to create/delete with every test case? Can we use multiple GSAs?

Reply:
The test case creates/deletes these bindings for two purposes:
1. Identify the minimal bindings required for each auth type (either gcpserviceaccount or k8sserviceaccount).
2. Add validations for the KNV2004 errors. There're actually two types of  permission errors: permissions not granted, and IAM bindings not created for the WI. As a side-effect, these validations align the testcases with the documentation. After these changes, we need to update the documentation for each invalid scenario.

Reply:
I'm still a bit concerned about the risk. I imagine the test could verify the various error modes by switching between multiple GSAs. Is there a reason this wouldn't work?

Reply:
Switching between multiple GSAs doesn't seem to be a normal user flow and will trigger a Pod restart.
By using the same GSA with IAM bindings created/deleted, this PR identifies the repulling issue in oci-sync, and helm-sync requires a more frequent re-pulling period.

Reply:
It just occurred to me a larger issue with this is that this strategy won't work when multiple tests are running against the same project in parallel, as they do in CI. I expect this would cause a lot of test flakiness due to tests stepping on each other

Reply:
That's right. I was testing all the tests in sequence.
What about creating a unique GSA for each test, and then delete it after?

Reply:
Is it necessary that we test re-polling in an e2e test? Perhaps a unit test would suffice.

Creating a unique GSA for each test could probably work for CI.

Reply:
Hmm, re-polling test can be done in a unit test, but there is probably no way to distinguish the following KNV2004 errors:
1. missing IAM binding between GSA and KSA: https://github.com/GoogleContainerTools/kpt-config-sync/blob/d4d2f694645ec092f7dcb44a582b7ec3ebf107e4/e2e/testcases/workload_identity_test.go#L231
2. missing IAM permission for GSA: https://github.com/GoogleContainerTools/kpt-config-sync/blob/d4d2f694645ec092f7dcb44a582b7ec3ebf107e4/e2e/testcases/workload_identity_test.go#L250

It maybe okay as we will recommend using BYOID over GSA for simplicity.

> Creating a unique GSA for each test could probably work for CI.

It works with GSA, but not BYOId unless we create a unique RSync for each test.

Given the above reasons, let me change it back by assuming all permissions are pre-granted.

Reply:
The validation for different permission errors can be extracted to a separate PR.

This PR is only for fixing AR authenticator with retries.
It would be tricky to add a unit test because we will need to verify the main process won't exit abnormally (the previous behavior) when it fails to get a valid authenticator.

Reply:
This is related to BYOID, but not a requirement.
The issue only happens when the permission is missing initially, and then granted later. The workaround is to restart the Pod.

Given that. it doesn't need to be cherrypicked to the 1.17 branch, but can wait until 1.18.0.